### PR TITLE
Move NameRef classification logic out of reference_definition

### PR DESCRIPTION
**Pull request by lnicola**

### CODE REVIEW

File: crates/ra_ide_api/src/goto_definition.rs [hunk 1]
```diff
@@ -48,87 +55,24 @@ pub(crate) fn reference_definition(
...
-                    return Exact(NavigationTarget::from_field(db, field));
-                }
+    match classify_name_ref(db, file_id, name_ref) {
+        Some(Method(func)) => return Exact(NavigationTarget::from_function(db, func)),
+        Some(Macro(mac)) => return Exact(NavigationTarget::from_macro_def(db, mac)),
+        Some(FieldAccess(field)) => return Exact(NavigationTarget::from_field(db, field)),
+        Some(AssocItem(assoc)) => return Exact(NavigationTarget::from_impl_item(db, assoc)),
+        Some(Def(def)) => return Exact(NavigationTarget::from_def(db, def)),
+        Some(SelfType(ty)) => {
+            if let Some((def_id, _)) = ty.as_adt() {
```
Comment (lnicola) [inline crates/ra_ide_api/src/goto_definition.rs]: Should we handle this in `classify_name_ref`?

File: crates/ra_ide_api/src/name_ref_kind.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+    }
+
+    // General case, a path or a local:
+    if let Some(path) = name_ref.syntax().ancestors().find_map(ast::Path::cast) {
+        if let Some(resolved) = analyzer.resolve_path(db, path) {
+            return match resolved {
+                hir::PathResolution::Def(def) => Some(Def(def)),
+                hir::PathResolution::LocalBinding(Either::A(pat)) => Some(Pat(pat)),
```
Comment (lnicola) [inline crates/ra_ide_api/src/name_ref_kind.rs]: I split these two cases (patterns and `self`), is that a bad idea?

File: crates/ra_ide_api/src/name_ref_kind.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                hir::PathResolution::LocalBinding(Either::B(par)) => Some(SelfParam(par)),
+                hir::PathResolution::GenericParam(par) => {
+                    // FIXME: get generic param def
+                    Some(GenericParam(par))
+                }
+                hir::PathResolution::Macro(def) => Some(Macro(def)),
+                hir::PathResolution::SelfType(impl_block) => {
+                    let ty = impl_block.target_ty(db);
```
Comment (lnicola) [inline crates/ra_ide_api/src/name_ref_kind.rs]: Maybe leave this for the caller to do?

File: crates/ra_ide_api/src/name_ref_kind.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    GenericParam(u32),
+}
+
+pub(crate) fn classify_name_ref(
+    db: &RootDatabase,
+    file_id: FileId,
+    name_ref: &ast::NameRef,
+) -> Option<NameRefKind> {
+    use NameRefKind::*;
+    let analyzer = hir::SourceAnalyzer::new(db, file_id, name_ref.syntax(), None);
```
Comment (lnicola) [inline crates/ra_ide_api/src/name_ref_kind.rs]: Can `SourceAnalyzer`s be reused somehow?

**Comment by lnicola** [comment]
Where did Travis go?

File: crates/ra_ide_api/src/name_ref_kind.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    FieldAccess(hir::StructField),
+    AssocItem(hir::ImplItem),
+    Def(hir::ModuleDef),
+    SelfType(hir::Ty),
+    Pat(AstPtr<ast::Pat>),
+    SelfParam(AstPtr<ast::SelfParam>),
+    GenericParam(u32),
+}
+
+pub(crate) fn classify_name_ref(
```
Comment (lnicola) [inline crates/ra_ide_api/src/name_ref_kind.rs]: I should probably collapse the `None` into `NameRefKind`.

File: crates/ra_ide_api/src/name_ref_kind.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    // Check if it is a method
+    if let Some(method_call) = name_ref.syntax().parent().and_then(ast::MethodCallExpr::cast) {
+        tested_by!(goto_definition_works_for_methods);
+        if let Some(func) = analyzer.resolve_method_call(method_call) {
+            return Some(Method(func));
+        }
+    }
+
+    //it could be a macro call
```
Comment (kjeremy) [inline crates/ra_ide_api/src/name_ref_kind.rs]: Nit: fix the comment

File: crates/ra_ide_api/src/goto_definition.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_ide_api/src/goto_definition.rs]: I think the current approach is good!

File: crates/ra_ide_api/src/name_ref_kind.rs (hunk 5 above)
Comment (matklad) [reply crates/ra_ide_api/src/name_ref_kind.rs]: Hm, I think `Option<NameRefKind>` makes more sense, as it works with `?`

File: crates/ra_ide_api/src/name_ref_kind.rs (hunk 4 above)
Comment (matklad) [reply crates/ra_ide_api/src/name_ref_kind.rs]: Good question! In general, `SourceAnalyzer` should be reused on per-function basis. 

Let's pass it in as an argument to the function? If the caller needs to classify several refs, they can then re-use the same analyzer.

File: crates/ra_ide_api/src/name_ref_kind.rs (hunk 2 above)
Comment (matklad) [reply crates/ra_ide_api/src/name_ref_kind.rs]: I think this is much better than `Either`!

File: crates/ra_ide_api/src/name_ref_kind.rs (hunk 3 above)
Comment (matklad) [reply crates/ra_ide_api/src/name_ref_kind.rs]: I don't have a stong opinion here, but returning `ty` seems more convenient.

File: crates/ra_ide_api/src/name_ref_kind.rs (hunk 4 above)
Comment (lnicola) [reply crates/ra_ide_api/src/name_ref_kind.rs]: I was confused because `SourceAnalyzer::new` takes (and uses) the `&SyntaxNode` of as an argument.

Comment (matklad) [reply crates/ra_ide_api/src/name_ref_kind.rs]: Yeah, but the Node needn't be exactly the one you will be asking questions about. Any child node of the function would do.

The API is slightly confusing (nothing prevents you from passing a node from *another* function, and that would be an error), but it is in general not clear yet what is the best way to bridge semantic info from hir and source-level info from parse trees. `SourceAnalyzer` is definitely better than the thing we had before :-)

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**